docs/design/2022-03-09-optimize-gc-for-internal-transaction.md

```diff
@@ -0,0 +1,141 @@
+
+```go
+type SessionManager interface {
+	// Put the internal session pointer to the map in the SessionManager
+	StoreInternalSession(se interface{})
+	// Delete the internal session pointer from the map in the SessionManager
+	DeleteInternalSession(se interface{})
+	// Get all startTS of every transactions running in the current internal sessions
+	GetInternalSessionStartTSList() []uint64
+}
+```
+
+#### innerTxnStartTsBox 
+
+This design defines a global variable `globalInnerTxnTsBox` to manage internal transactions run by function `RunInNewTxn`. The map `innerTxnStartTsBox.innerTxnStartTsMap` stores all the internal transaction startTS. 
+
+```go
+var globalInnerTxnTsBox = innerTxnStartTsBox{
+	innerTSLock:        sync.Mutex{},
+	innerTxnStartTsMap: make(map[uint64]struct{}, 256),
+}
+
+type innerTxnStartTsBox struct {
+	innerTSLock        sync.Mutex
+	innerTxnStartTsMap map[uint64]struct{}
+}
+```
+
+### Functional design
+
+#### Store And Delete Internal Sessions
+
+TiDB gets an internal session from a session pool implemented in `(s *session) getInternalSession`, when the transaction finished,it puts the session to the pool. This design stores the internal session to session manger and deletes the internal session from session manager in the function  `(s *session) getInternalSession`. It calls the function `infosync.DeleteInternalSession()` to delete the internal session from session manager and calls the function `infosync.StoreInternalSession` to add the internal session to session manger.
+
+```go
+func (s *session) getInternalSession(execOption sqlexec.ExecOption) (*session, func(), error) {
+	tmp, err := s.sysSessionPool().Get()
+	se := tmp.(*session)
+	// Put the internal session to the map of SessionManager
+	infosync.StoreInternalSession(se)
+	return se, func() {
+		// Delete the internal session to the map of SessionManager
+		infosync.DeleteInternalSession(se)
+		s.sysSessionPool().Put(tmp)
+	}
+}
+```
+
+#### Store And Delete Internal Transactions
+
+The internal transactions here are executed by function `RunInNewTxn`. It stores the internal transaction startTS by function  `wrapStoreInterTxnTS()` and deletes the transaction startTS by function `wrapDeleteInterTxnTS()`.
+
+```go
+// RunInNewTxn will run the f in a new transaction environment.
+func RunInNewTxn(ctx context.Context, store Storage, retryable bool, f func(ctx context.Context, txn Transaction) error) error {
+	defer func() {
+		wrapDeleteInterTxnTS(originalTxnTS)
+	}()
+    for i := uint(0); i < maxRetryCnt; i++ {
+        txn, err = store.Begin()
+        		if i == 0 {
+			originalTxnTS = txn.StartTS()
+			wrapStoreInterTxnTS(originalTxnTS)
+		}
+		err = txn.Commit(ctx)
+        ...
+    }
+}
+```
+
+#### Calculate GC Safe Point
+
+Currently, TiDB calculates gc safe point in the function `(is *InfoSyncer) ReportMinStartTS`. This design add some code in the  function `ReportMinStartTS` to consider internal transactions when calculates gc safe point.
```

Comment:
I'd prefer saying `ReportMinStartTS` calculates the minimum startTS of ongoing transactions (except the ones that exceed the limit), and it's result is useful for calculating the safe point, which is gc_worker's work. The `ReportMinStartTS` doesn't calculate the safe point directly.